tracker/view/login.py

```diff
@@ -8,39 +8,105 @@
 
 from config import TRACKER_PASSWORD_LENGTH_MAX
 from config import TRACKER_PASSWORD_LENGTH_MIN
-from tracker import tracker
+from config import SSO_ENABLED, SSO_NEW_USER_DEFAULT_PASSWORD, SSO_ADMINISTRATOR_GROUP, SSO_REPORTER_GROUP, SSO_SECURITY_TEAM_GROUP, SSO_GUEST_GROUP
+from tracker import tracker, oauth
 from tracker.form import LoginForm
 from tracker.model.user import User
 from tracker.user import user_assign_new_token
 from tracker.user import user_invalidate
+from ..model.enum import UserRole
 
 
 @tracker.route('/login', methods=['GET', 'POST'])
 def login():
-    if current_user.is_authenticated:
-        return redirect(url_for('tracker.index'))
+    if SSO_ENABLED:
+        return redirect(url_for('tracker.list_user'))
+    else:
+        if current_user.is_authenticated:
+            return redirect(url_for('tracker.index'))
 
-    form = LoginForm()
-    if not form.validate_on_submit():
-        status_code = Unauthorized.code if form.is_submitted() else 200
-        return render_template('login.html',
-                               title='Login',
-                               form=form,
-                               User=User,
-                               password_length={'min': TRACKER_PASSWORD_LENGTH_MIN,
-                                                'max': TRACKER_PASSWORD_LENGTH_MAX}), status_code
-
-    user = user_assign_new_token(form.user)
-    user.is_authenticated = True
-    login_user(user)
-    return redirect(url_for('tracker.index'))
+        form = LoginForm()
+        if not form.validate_on_submit():
+            status_code = Unauthorized.code if form.is_submitted() else 200
+            return render_template('login.html',
+                                title='Login',
+                                form=form,
+                                User=User,
+                                password_length={'min': TRACKER_PASSWORD_LENGTH_MIN,
+                                                 'max': TRACKER_PASSWORD_LENGTH_MAX}), status_code
+
+        user = user_assign_new_token(form.user)
+        user.is_authenticated = True
+        login_user(user)
+        return redirect(url_for('tracker.index'))
 
 
 @tracker.route('/logout', methods=['GET', 'POST'])
 def logout():
+    # TODO clear SSO session
     if not current_user.is_authenticated:
         return redirect(url_for('tracker.index'))
 
     user_invalidate(current_user)
     logout_user()
     return redirect(url_for('tracker.index'))
+
+
+@tracker.route('/sso-auth')
+def sso_auth():
```

Comment:
This function might be missing updating the user groups, when they are changed in the identity provider.

In archweb we also store the `Open ID Connect Subject Identifier` to identify the OIDC account to the local account without relying on the email address

Reply:
Ref: https://openid.net/specs/openid-connect-core-1_0.html#ClaimStability

Reply:
> This function might be missing updating the user groups, when they are changed in the identity provider.
I think it does that, because the groups are transmitted on every login. Even with the user existing, it updates the user groups. Or am I misunderstanding you here?

> In archweb we also store the Open ID Connect Subject Identifier to identify the OIDC account to the local account without relying on the email address
OK, so my takeaway is that I should use the subject identifier as ID for the user instead of his email?
The thing is, that in the tracker DB three attributes of the user must be unique: ID, email and name.
I am happy to make the change, but I think it's preferrable to stay consistent with the email/password authentication method which also happens based on email.

Reply:
> OK, so my takeaway is that I should use the subject identifier as ID for the user instead of his email?

You should try to match against the `sub` first and fallback to using the email address. So the email address is only used for the initial linking. That of course means that we need to store `sub` in the db.

Reply:
Okay, fixed this in a2535c6 
I need some guidance on how to add a migration for this, because it looks like this is a custom solution. Could you help me there? Luckily it's just an extra column, so it could be done by hand easily in the worst case

Reply:
Looks mostly good, a few things to fix, to make it secure:
1. The `email_verified` claim should be checked before trusting the email address
2. If the user already exists, the `email` should be updated on login
3. The user should only be matched on `email` if `idp_id` is `None` for the returned user

Reply:
@jelly is the migration wizard I think (at least more than me ;))

Reply:
> The email_verified claim should be checked before trusting the email address
Do you mean that if matching the user via email address, the `email_verified` flag should be taken into account?
Or that upon creating the user this flag should be set to true?

> If the user already exists, the email should be updated on login
Good point, added that in 9df3befc

> The user should only be matched on email if idp_id is None for the returned user
That is already the case. Only if the user can't be matched based on the `idp_id`, the email is taken into account

Reply:
> Do you mean that if matching the user via email address, the `email_verified` flag should be taken into account?

Yeh, the claim is documented here: https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims
>  True if the End-User's e-mail address has been verified; otherwise false. When this Claim Value is true, this means that the OP took affirmative steps to ensure that this e-mail address was controlled by the End-User at the time the verification was performed. The means by which an e-mail address is verified is context-specific, and dependent upon the trust framework or contractual agreements within which the parties are operating.

So if the claim is false just stop the authentication process and `return redirect(url_for('tracker.index'))`.

> That is already the case. Only if the user can't be matched based on the `idp_id`, the email is taken into account

Correct, but if user A changes his mail address, user B could change to user A old mail address and login as user A. So we would need something like:
```python
user = db.get(User, idp_id=parsed_token.get('sub'))
if user == None:
    user = db.get(User, email=user_email_idp)
    if user.idp_id != None:
        return redirect(url_for('tracker.index'))
```
(not a Python programmer, so perhaps it can be done in a better way)

Reply:
Thanks for the help! For some reason I thought I remembered that `email_verified` was on the user model. I didn't think of the claim. Added that just now in 0187a5d.
To sum up, that's now three reasons when a successful authentication in the IDP can be denied here:
1. no groups are given
2. trying to impersonate somebody else who has an IDP account
3. user has not confirmed his email address yet

Reply:
Good work :)

I think this:
```python
    if user.idp_id != None:
        print("SSO error: user sub {} tried to authenticate as {}".format(user_sub, user.email))
        return redirect(url_for('tracker.index'))
```
should be:
```python
    if user.idp_id != None and user.idp_id != user_sub:
        print("SSO error: user sub {} tried to authenticate as {}".format(user_sub, user.email))
        return redirect(url_for('tracker.index'))
```

Reply:
Hehe, I think that case can never happen because if they are indeed the same, the first query would have gotten the user by the `idp_id` straight away

Reply:
Would `user.idp_id` be `None` in that case?

Reply:
Yes, I set the default value to `None` in the model

Reply:
I don't get it. Are you saying that `db.get(User, idp_id=user_sub)` will never return a user where `idp_id` isn't `None`?

Basically I'm asking why `if user.idp_id != None:` wouldn't be true if the user has logged in before.

Reply:
You were totally right. I was thinking from one end, and you from the other :D
If you check the most recent changes in c811039, you'll see that the part that we were circling around has changed quite a bit.